<h1><center>Laboratorio 2: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Diego Irarrázaval

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1:
- Nombre de alumno 2:


### **Link de repositorio de GitHub:** `http://....`

### Indice 

1. [Temas a tratar](#Temas-a-tratar:)
3. [Descripcción del laboratorio](#Descripción-del-laboratorio.)
4. [Desarrollo](#Desarrollo)

# Temas a tratar

- Manejo de DataFrames con la libreria `pandas`.
- Visualizaciones interactivas de los datos con `plotly`.

## Reglas:

- Fecha de entrega: TBD
- **Grupos de 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer material y segmentos de código del curso que estimen conveniente.

### Objetivos principales del laboratorio

- Aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas`.
- Utilizar `plotly` para obtener información gráfica del dataset.
- Aplicar el **Análisis Exploratorio de Datos** a un caso en particular.


El laboratorio deberá ser desarrollado **sin el uso indiscriminado de nativos de python** (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

#Importamos librerias utiles 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.express as px

# 1. Rendimiento en Estudiantes 📚


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

Para este laboratorio deberán realizar un Análisis Exploratorio de datos sobre el conjunto ```students_grades.csv```, el cual contiene una caracterización sobre el rendimiento y otros atributos de cada alumno de la *Universidad de la Cachaña* .

In [ ]:
# Usar solamente si utilizan Collab.
from google.colab import drive
drive.mount("/content/drive")


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [133]:
dir = '/Users/matiassepulveda/Documents/Semestre Otoño 2021/Laboratorio de Ciencia de Datos/Laboratorio 2/202104231153C65226A3407DEBA6__Dataset_lab2/'
df_grades = pd.read_csv(dir+'students_grades.csv')

## 1.1 Revisión de datos [1 punto]

<p align="center">
  <img src="https://media2.giphy.com/media/4FQMuOKR6zQRO/giphy.gif?cid=ecf05e47ii9lhu9tc9tb31d6lafrysp2zyf6u8w3y348mk8t&rid=giphy.gif&ct=g">
</p>



El Rector *Don Caguayo* le comenta que actualmente la Universidad posee un total de **1000** alumnos inscritos. Sin embargo, al revisar su *Excel*, no le cabe en la cabeza que hayan más de **1000** alumnos en la universidad. 

Consternado, el rector cree que se han realizado malas prácticas al inscribir tanto los registros de las notas como las notas como tal y les solicita a ustedes (aka, los expertos) que revisen los datos. 

Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**.

Cree una función que le permita revisar si las diferentes variables que componen el dataset son coherentes y si es posible observar anomalías en estas.

> ¿Sera necesario limpiar los datos?, comente esto en la celda ubicada debajo de **justificación**. (agregar si están bien el rango de notas)

**To-Do**

Crear la función que `analisis_exploratorio` que:
- [ ] Imprima un extracto del Dataframe (usar `display` en vez de `print` para los DataFrames).
- [ ] Imprima las dimensiones del DataFrame.
- [ ] Describa los datos numéricos a través de ```.describe()```.
- [ ] Imprima el total de filas repetidas (hint: Buscar el método```duplicated()```.).
- [ ] Imprima la cantidad de valores nulos (`isna()`).
- [ ] Imprima la cantidad de valores unicos.

Además, genere visualizaciones (fuera de la función) que:
- [ ] Muestre la distribución de las variables numericas.

Por último, 

- [ ] Justifique si será necesario limpiar los datos y aplique los análisis univariado y/o bivariado vistos en clase.


In [134]:
def analisis_preliminar_datos(df_grades):
    display(df_grades[0:5])
    print('Dimensiones del set de datos (Filas x Columnas): ',df_grades.shape)
    print('')
    print('Descripción de los datos númericos: ')
    print(df_grades.describe())      
    duplicados = sum(df_grades.duplicated())
    print('')
    print('Cantidad de filas duplicadas: ', duplicados)
    print('')
    print('Cantidad total de valores NAs: ', sum(df_grades.isna().sum()))
    print('')
    print('Cantidad total de valores únicos por columna: ')
    print('')
    print(df_grades.nunique())
    print('')
    print('Cantidad total de filas únicas: ', len(df_grades) - duplicados)

In [135]:
analisis_preliminar_datos(df_grades)

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0

Dimensiones del set de datos (Filas x Columnas):  (1400, 9)

Descripción de los datos númericos: 
        math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000

Cantidad de filas duplicadas:  3

Cantidad total de valores NAs:  127

Cantidad total de valores únicos por columna: 

names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64

Can

> Utilice alguna visualización (gráfique) para observar el comportamiento de los datos numericos. ¿Estan todos los valores correctos?

In [136]:
# Se procede a visualizar las distribuciones de las variables numericas, se escribe en una misma linea para no
# ensuciar el código

#px.histogram(df_grades, x= 'reading score', title='Análisis de la variable reading score', marginal='violin')
#px.histogram(df_grades, x= 'math score', title='Análisis de la variable math score', marginal='violin')
px.histogram(df_grades, x= 'writing score', title='Análisis de la variable writing score', marginal='violin')

**Justificación:**

```
Se puede ver en las distribuciones anteriores y los resultados de la función analisis_preliminar_datos() que existen valores que exceden el umbral posible de 100 como porcentaje de logro, es más se puede ver que los valores anteriores que caben dentro del umbral 0 a 100 tienen un comportamiento bastante normal en comparación a estos datos que son cuestionables.

Otro aspecto importante, es notar ya que vemos un comportamiento bastante normal en los datos que se encuentran en el umbral, la mediana no debería estar tan alejada del promedio, pero debido a estos otros datos cuestionables el promedio se aleja bastante.

```

## 1.2 Limpieza de Datos [1 punto]

<p align="center">
  <img src="https://media3.giphy.com/media/c09yGbnSyaFRS/giphy.gif?cid=ecf05e47d0ejjjnpfj0ywcbwz7zk7sw2rfdpis890gmdpvj2&rid=giphy.gif&ct=g">
</p>


En base a lo realizado anteriormente, tome la decisión (si es necesario) de limpiar el dataset. Asuma los supuestos que estime conveniente para limpiar el DataFrame y cree una función que le permita limpiar los datos, si estima que no es necesario realizar esto, omita esta parte.

**To-Do:**

- [ ] Decidir si realizar o no la limpieza.
- [ ] Crear función que nos permita limpiar el dataset.
- [ ] Justifique las asunciones realizadas para limpiar los datos.

**Hint**

En el caso que decida limpiar el dataset, puede que le sean de utilidad las funciones ```.drop_duplicates()``` y ```.dropna()```.


In [137]:
def limpiar_dataset(df_grades):
    # Primero vamos a filtrar para mantener sólo los valores de las pruebas que sean menores a 100, ya que no es factible 
    # tener un puntaje mayor de ese umbral
    df_grades = df_grades[(df_grades['math score'] <= 100) & (df_grades['reading score']<=100) & (df_grades['writing score']<=100)]
    # Luego vamos a filtrar los nombres únicos ya que no es posible que un alumno haya dado más de una vez el mismo examen.
    df_grades = df_grades.drop_duplicates('names')
    # Finalmente se decide no filtrar los valores NA, a continuación se explica por qué.
    return df_grades

In [138]:
# Ejecutar antes para luego probar analisis_preliminar_datos
df_grades = limpiar_dataset(df_grades)

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [139]:
analisis_preliminar_datos(df_grades)

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
3  standard               completed        80.0           78.0           81.0  
5  standard                    none        73.0           66.0           63.0  
6  standard               completed        81.0           75.0           76.0  
7  standard               completed        94.0           86.0           87.0

Dimensiones del set de datos (Filas x Columnas):  (1000, 9)

Descripción de los datos númericos: 
       math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

Cantidad de filas duplicadas:  0

Cantidad total de valores NAs:  0

Cantidad total de valores únicos por columna: 

names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64

Cantidad total

**Justificación:**

```
Retomando lo anteriormente detallado, se decide primero filtrar los filas de las pruebas que exceden el umbral específicado de 100 (antes en la distribución se ve que no bajan de 0), ya que según las indicaciones esto no es posible ya que es un porcentaje de logro.

Luego, se decide filtrar los nombres duplicados, ya que no es posible que una persona haya rendido más de una vez estas pruebas.

Finalmente, se decide no eliminar NA en caso de que hayan quedado en el set de datos debido a que esta información puede ser valiosa en caso de que pueda ser un indicativo de que un determinado sector de los estudiantes no haya rendido las pruebas por algún motivo en especial y esto puede ser explicado mediante los datos.
```

## 1.3 Análisis exploratorio de datos



<p align="center">
  <img src="https://media4.giphy.com/media/ko7twHhomhk8E/giphy.gif?cid=ecf05e47xmgfz8ftaom8saa4cwktxpb0fkui9yk3nakb4ff1&rid=giphy.gif&ct=g">
</p>


### 1.3.1 Transformación de notas a "Nota Chilena" [0.5 punto]

En la ultima reunion con Don Caguayo este les comenta es complejo para el y su equipo entender la graduación de notas de 0 a 100. Por ende, le comenta que desde hace harto tiempo tienen la intención de *trnasformar* todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree un función que le permita pasar a la escala Chilena las notas que se escapan del rango [1,7]. 

(No ocupe ningún tipo de ciclo).

**To-Do**

- [ ] Tranforme los valores a la escala chilena.
- [ ] Asigne los valores transformados al DataFrame.

In [140]:
def transformar_notas_dataset(df_grades):
    math_nota_chilena = df_grades['math score']*0.06 + 1
    reading_nota_chilena = df_grades['reading score']*0.06 + 1
    writing_nota_chilena = df_grades['writing score']*0.06 + 1
    df_grades['math score'] = math_nota_chilena 
    df_grades['reading score'] = reading_nota_chilena
    df_grades['writing score'] = writing_nota_chilena
    return df_grades

In [141]:
# Ejecutar esta celda sólo 1 vez antes de continuar
df_grades = transformar_notas_dataset(df_grades)

### 1.3.2 Obtención de promedio de alumnos [0.5 puntos]

Con todas las notas adjuntadas, la prestigiosa *Universidad De la Cachaña* necesita obtener el promedio general de cada uno de los alumnos (Grade Point Average - GPA). Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del DataFrame original.

**To-Do:**
- [ ] Obtenga el promedio de cada uno de los alumnos.
- [ ] Anexe esta información en el DataFrame original.


In [143]:
GPA = (df_grades['math score'] + df_grades['reading score'] + df_grades['writing score'])/3
df_grades['GPA'] = GPA
df_grades.head(5)

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
3  5.78  
5  5.04  
6  5.64  
7  6.34

### 1.3.3 Reprobados [1 punto]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena.

Cree una función que permita obtener dos Dataframes, el primero debe poseer solamente los alumnos que han reprobado en la prestigiosa universidad y el segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.

**To-Do:**

- [ ] Obtenga una máscara DataFrame con los alumnos que reprobaron el año.
- [ ] Reemplaze todos los GPA reprobados con una **R**.

In [144]:
reprobados = df_grades[df_grades['GPA'] < 4.0]
df_grades_R = df_grades.copy()
df_grades_R['GPA'] = df_grades['GPA'].replace([reprobados['GPA']] , 'R')

In [145]:
df_grades_R

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score   GPA  
0              3.76     R  
3              5.86  5.78  
5              4.78  5.04  
6              5.56  5.64  
7              6.22  6.34  
...             ...   ...  
1392           5.26   5.3  
1394           3.16     R  
1396           5.86  5.42  
1397           4.72  4.54  
1399           5.56  5.48  

[1000 rows x 10 columns]

### 1.3.4 Información por Categorías [1 punto]

Motivado por saber más sobre los datos, el Rector les solicita que genere una función que le permita comparar los promedios de diferentes **atributos categóricos** del DataFrame. Para esto su función deberá tener como entrada el DataFrame con todos los promedios (sin agregar las R's) y calcular el promedio **agrupando por etnias y el nivel educacional de los padres**.

**To-Do:**

- [ ] Hacer una función que nos permita obtener la comparación de promedios para dos atributos categóricos.
- [ ] Cada uno de los grupos expuestos deben exponer el porcentaje de alumnos que representan en el colegio.
- [ ] Comente lo que observa de los resultados.
- [ ] Visualice los datos generados con [`px.sunburst`](https://plotly.com/python/sunburst-charts/ ).


In [149]:
def compara_promedios(df_grades):
    df_grades_agrupado = df_grades.groupby(['race/ethnicity', 'parental level of education']).mean().round(2)
    count = df_grades.groupby(['race/ethnicity', 'parental level of education']).count()['names']
    df_grades_agrupado['percentaje %'] = 100*count/len(df_grades)
    return df_grades_agrupado

In [150]:
compara_promedios(df_grades)

math score  reading score  \
race/ethnicity parental level of education                              
group A        associate's degree                 4.66           5.02   
               bachelor's degree                  5.03           5.08   
               high school                        4.63           4.77   
               master's degree                    4.46           4.88   
               some college                       4.83           4.95   
               some high school                   4.54           4.72   
group B        associate's degree                 4.97           5.18   
               bachelor's degree                  5.16           5.38   
               high school                        4.59           4.81   
               master's degree                    5.03           5.81   
               some college                       4.79           4.95   
               some high school                   4.71           4.99   
group C        associate's degree                 5.00           5.27   
               bachelor's degree                  5.09           5.54   
               high school                        4.65           4.87   
               master's degree                    5.02           5.23   
               some college                       4.91           5.17   
               some high school                   4.63           4.94   
group D        associate's degree                 5.06           5.23   
               bachelor's degree                  5.05           5.21   
               high school                        4.77           4.86   
               master's degree                    5.35           5.63   
               some college                       5.12           5.25   
               some high school                   5.01           5.20   
group E        associate's degree                 5.49           5.43   
               bachelor's degree                  5.59           5.49   
               high school                        5.25           5.22   
               master's degree                    5.48           5.93   
               some college                       5.43           5.36   
               some high school                   5.33           5.17   

                                            writing score   GPA  percentaje %  
race/ethnicity parental level of education                                     
group A        associate's degree                    4.81  4.83           1.4  
               bachelor's degree                     5.10  5.07           1.2  
               high school                           4.63  4.68           1.8  
               master's degree                       5.06  4.80           0.3  
               some college                          4.90  4.89           1.8  
               some high school                      4.51  4.59           2.4  
group B        associate's degree                    5.09  5.08           4.1  
               bachelor's degree                     5.30  5.28           2.0  
               high school                           4.68  4.69           4.8  
               master's degree                       5.63  5.49           0.6  
               some college                          4.85  4.86           3.7  
               some high school                      4.88  4.86           3.8  
group C        associate's degree                    5.22  5.16           7.8  
               bachelor's degree                     5.55  5.39           4.0  
               high school                           4.70  4.74           6.4  
               master's degree                       5.17  5.14           1.9  
               some college                          5.13  5.07           6.9  
               some high school                      4.80  4.79           4.9  
group D        associate's degree                    5.19  5.16           5.0  
               bachelor's d

**Respuesta Esperada:**

| race/ethnicity | parental level of education | math score | reading score | writing score |  GPA | percentaje % |
|---------------:|----------------------------:|-----------:|--------------:|--------------:|-----:|-------------:|
|     group A    |      associate's degree     |       4.66 |          5.02 |          4.81 | 4.83 |          1.4 |
|                |      bachelor's degree      |       5.03 |          5.08 |          5.10 | 5.07 |          1.2 |
|                |         high school         |       4.63 |          4.77 |          4.63 | 4.68 |          1.8 |
|                |       master's degree       |       4.46 |          4.88 |          5.06 | 4.80 |          0.3 |
|                |         some college        |       4.83 |          4.95 |          4.90 | 4.89 |          1.8 |
|                |       some high school      |       4.54 |          4.72 |          4.51 | 4.59 |          2.4 |
|     group B    |      associate's degree     |       4.97 |          5.18 |          5.09 | 5.08 |          4.1 |
|                |      bachelor's degree      |       5.16 |          5.38 |          5.30 | 5.28 |          2.0 |
|                |         high school         |       4.59 |          4.81 |          4.68 | 4.69 |          4.8 |
|                |       master's degree       |       5.03 |          5.81 |          5.63 | 5.49 |          0.6 |
|                |         some college        |       4.79 |          4.95 |          4.85 | 4.86 |          3.7 |
|                |       some high school      |       4.71 |          4.99 |          4.88 | 4.86 |          3.8 |
|     group C    |      associate's degree     |       5.00 |          5.27 |          5.22 | 5.16 |          7.8 |
|                |      bachelor's degree      |       5.09 |          5.54 |          5.55 | 5.39 |          4.0 |
|                |         high school         |       4.65 |          4.87 |          4.70 | 4.74 |          6.4 |
|                |       master's degree       |       5.02 |          5.23 |          5.17 | 5.14 |          1.9 |
|                |         some college        |       4.91 |          5.17 |          5.13 | 5.07 |          6.9 |
|                |       some high school      |       4.63 |          4.94 |          4.80 | 4.79 |          4.9 |
|     group D    |      associate's degree     |       5.06 |          5.23 |          5.19 | 5.16 |          5.0 |
|                |      bachelor's degree      |       5.05 |          5.21 |          5.31 | 5.19 |          2.8 |
|                |         high school         |       4.77 |          4.86 |          4.79 | 4.81 |          4.4 |
|                |       master's degree       |       5.35 |          5.63 |          5.78 | 5.59 |          2.3 |
|                |         some college        |       5.12 |          5.25 |          5.30 | 5.23 |          6.7 |
|                |       some high school      |       5.01 |          5.20 |          5.15 | 5.12 |          5.0 |
|     group E    |      associate's degree     |       5.49 |          5.43 |          5.39 | 5.44 |          3.9 |
|                |      bachelor's degree      |       5.59 |          5.49 |          5.52 | 5.54 |          1.8 |
|                |         high school         |       5.25 |          5.22 |          5.05 | 5.17 |          2.2 |
|                |       master's degree       |       5.48 |          5.93 |          5.83 | 5.74 |          0.8 |
|                |         some college        |       5.43 |          5.36 |          5.21 | 5.33 |          3.5 |
|                |       some high school      |       5.33 |          5.17 |          4.99 | 5.16 |          1.8 |

> Visualice utilizando [`px.sunburst`](https://plotly.com/python/sunburst-charts/ ) el DataFrame obtenido en esta sección. 

In [182]:
df =  compara_promedios(df_grades).reset_index()
px.sunburst(df, path=['race/ethnicity', 'parental level of education'], 
                  values='percentaje %', color = 'GPA')

> Comente los resultados obtenidos. ¿Hay alguna diferencia entre los grupos graficados?, ¿Es útil esta visualización?

**Justifique:**

```
Efectivamente hay diferencia entre los grupos graficados, en el primer nivel de la variable 'race/ethnicity', se pueden apreciar resultados diferentes entre los grupos, por ejemplo el grupo E tiene un promedio de GPA de 5.36 y el grupo A tiene un promedio de 4.78, lo cuál es bastante diferencia, es más el grupo A representa un 8.9% de los estudiantes y el grupo E un 14%, aunque el grupo que tiene un porcentaje mayor es el C, con un 31.9 de los estudiantes.
Ahora, si nos vamos a la segunda variable 'parental level of education', nuevamente se evidencian resultados diferentes entre los alumnos que pertenecen a uno y otro.

Es muy útil la visualización, rápidamente de manera interactiva muestra las diferencias entre los diferentes grupos, y los diferentes niveles de educación de sus padres, además la barra de colores al lado nos favorece para buscar rápidamente los que tienen menores o mayores GPA e incluso nos da una idea de la distribución de los valores de GPA.
```

### 1.3.5 Combinar Dataset [1 punto]

Al notificarle los resultados obtenidos en la etapa de descubrimiento a Don Caguayo, este les informa que ha encontrado un Excel con nueva información, en esta se tienen los valores de dos asignaturas y desea que la unan con el dataset original (```students_grades.csv```).

Busque la forma de unir ambos DataFrames, para esto deberá anexar las columnas de ```history score``` y ```science score``` al final del DataFrame original.

**To-Do**

- [ ] Unir DataFrame ```students_grades.csv``` y ```other_grades.csv```.

In [207]:
df_other_grades = pd.read_csv(dir+'other_grades.csv')
df_grades = pd.read_csv(dir+'students_grades.csv')

In [208]:
pd.merge(left=df_grades, right= df_other_grades, on='names', how="inner")

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1        Rita Courtney  female        group B            some high school   
2     Charles Linstrom    male        group A           bachelor's degree   
3          Brian Young    male        group C            some high school   
4          Brian Young    male        group C            some high school   
...                ...     ...            ...                         ...   
1242     Richard Young    male        group D                 high school   
1243     Wanda Russell  female        group B                 high school   
1244    Marina Zeigler  female        group C           bachelor's degree   
1245     Laurie Carter  female        group B            some high school   
1246      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       193.0          193.0   
2         standard               completed        80.0           78.0   
3         standard                    none        73.0           66.0   
4         standard                    none         NaN            NaN   
...            ...                     ...         ...            ...   
1242      standard                    none        69.0           75.0   
1243  free/reduced               completed        23.0           44.0   
1244  free/reduced               completed        66.0           74.0   
1245      standard               completed        54.0           61.0   
1246      standard               completed        68.0           80.0   

      writing score  science score  history score  
0              46.0             25            3.8  
1             193.0             25            3.8  
2              81.0             56            5.2  
3              63.0             58            3.8  
4               NaN             58            3.8  
...             ...            ...            ...  
1242           71.0             62            4.2  
1243           36.0             16            3.6  
1244           81.0             59            4.1  
1245           62.0             43            4.3  
1246           76.0             61            5.3  

[1247 rows x 11 columns]

> Explique (pensando en una operación de conjuntos) que hace la operación de combinación por defecto (`inner`).

La operación inner lo que hace es unir los elementos que se encuentran en ambas tablas, descartando el resto, es decir si tenemos un alumno que se encuentra en df_grades, pero no en df_other_grades, se descarta. Además, si un alumno se encuentra dos veces como pudimos ver en la primera parte de este laboratorio, los valores de las columnas nuevas que se encuentran en df_other_grades se copian para ambos, manteniendo ambas filas.

> Explique que hubiese sucedido si aplicaramos los siguientes merge en las variables anteriores:

- `left` 

Entrega sólo los elementos que se combinaron correctamente del primer set de datos que le entregamos, df_grades, es decir devuelve todo lo de ese dataset, y añade cuando puede (existe la llave) los valores de las columnas nuevas.

- `right`

Es análogo a Left pero esta vez se mantienen los elementos del set de datos que le entregamos en la posición right= de pd.merge.

- `outer`

Esta vez combina todos los elementos posibles (que tengan la misma llave en ambos set de datos) y conserva todos los elementos pertenecientes a ambos dataframes.


**Justificación:**

```
Hay diferentes operaciones que nos entregarían resultados similares o de buena manera, pero en especial para no perder información bajo la idea de la primera pregunta que indica que los NA igualmente nos entregan valor, utilizaría outer, para luego filtrar los que están fuera de rango, 

```

## 1.4 Más visualizaciones [Bonus] [1 punto]


**Desde aquí parte el bonus para el lab.**


<p align="center">
  <img src="https://media4.giphy.com/media/xT5LMWNOjGqJzUfyve/giphy.gif?cid=ecf05e47phkhvg6vadqvcoobvxwj9bzxxk38jjvkcc6eak6t&rid=giphy.gif&ct=g">
</p>




### 1.4.1 Rendimiento por Asignatura

Con el DataFrame de rendimiento para cada uno de los alumnos grafique como distribuyen los datos para cada uno de los puntajes obtenidos en cada una de las asignaturas.

Para realizar esta tarea se les recomienda que utilicen el comando ```px.histogram``` de plotly. Para esto puede ser útil procesar los datos con un melt, agrupando en una columna las asignaturas y en otra columna la nota obtenida en dicha asignatura.

Ejemplo del melt:

|    Variable   | Value |
|:-------------:|:-----:|
| math score    | 6.4   |
| math score    | 5.2   |
| ...           | ...   |
| writing score | 4.5   |

**To-Do:**

- [ ] Utilice `plotly` para visualizar el rendimiento para los diferentes grupos que componen el dataset.
- [ ] Responder a la pregunta, ¿Cuál es la asignatura que más les cuesta a los alumnos de la universidad de la Cachaña?


In [ ]:
#### Codigo Aquí ####

> ¿Cuál es la asignatura que más les cuesta a los alumnos de la universidad de la Cachaña?

**Comente:**

```
Escriba su respuesta aquí
```

### 1.4.2 Otras Visualizaciones

Genere dos visualizaciones extras que encuentre interesantes con estos datos. Agrupe los atributos que estime convenientes.


**To-Do:**
- [ ] Generar dos nuevas visualizaciones con los datos.

In [ ]:
#### Codigo Aquí ####

# Conclusión
Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>

<br>
<center>
<img src="https://i.kym-cdn.com/photos/images/original/001/194/195/b18.png" width=100 height=50 />
</center>
<br>